service/history/queuev2/virtual_slice_test.go

```diff
@@ -0,0 +1,1562 @@
+					},
+					NextTaskKey: persistence.NewImmediateTaskKey(2),
+				},
+				{
+					Range: Range{
+						InclusiveMinTaskKey: persistence.NewImmediateTaskKey(5),
+						ExclusiveMaxTaskKey: persistence.NewImmediateTaskKey(6),
+					},
+					NextTaskKey: persistence.NewImmediateTaskKey(5),
+				},
+			},
+		},
+		{
+			name: "Case 3: [a,b,x,y,c,z] - Overlapping ranges with interleaved keys",
+			left: &GetTaskProgress{
+				Range: Range{
+					InclusiveMinTaskKey: persistence.NewImmediateTaskKey(1),
+					ExclusiveMaxTaskKey: persistence.NewImmediateTaskKey(5),
+				},
+				NextTaskKey: persistence.NewImmediateTaskKey(2),
+			},
+			right: &GetTaskProgress{
+				Range: Range{
+					InclusiveMinTaskKey: persistence.NewImmediateTaskKey(3),
+					ExclusiveMaxTaskKey: persistence.NewImmediateTaskKey(6),
+				},
+				NextTaskKey: persistence.NewImmediateTaskKey(4),
+			},
+			expected: []*GetTaskProgress{
+				{
+					Range: Range{
+						InclusiveMinTaskKey: persistence.NewImmediateTaskKey(2),
+						ExclusiveMaxTaskKey: persistence.NewImmediateTaskKey(3),
+					},
+					NextTaskKey: persistence.NewImmediateTaskKey(2),
+				},
+				{
+					Range: Range{
+						InclusiveMinTaskKey: persistence.NewImmediateTaskKey(4),
+						ExclusiveMaxTaskKey: persistence.NewImmediateTaskKey(6),
+					},
+					NextTaskKey: persistence.NewImmediateTaskKey(4),
+				},
+			},
+		},
+		{
+			name: "Case 4: [a,b,x,y,z,c] - Overlapping ranges with right extending beyond",
+			left: &GetTaskProgress{
+				Range: Range{
+					InclusiveMinTaskKey: persistence.NewImmediateTaskKey(1),
+					ExclusiveMaxTaskKey: persistence.NewImmediateTaskKey(6),
+				},
+				NextTaskKey: persistence.NewImmediateTaskKey(2),
+			},
+			right: &GetTaskProgress{
+				Range: Range{
+					InclusiveMinTaskKey: persistence.NewImmediateTaskKey(3),
+					ExclusiveMaxTaskKey: persistence.NewImmediateTaskKey(5),
+				},
+				NextTaskKey: persistence.NewImmediateTaskKey(4),
+			},
+			expected: []*GetTaskProgress{
+				{
+					Range: Range{
+						InclusiveMinTaskKey: persistence.NewImmediateTaskKey(2),
+						ExclusiveMaxTaskKey: persistence.NewImmediateTaskKey(3),
+					},
+					NextTaskKey: persistence.NewImmediateTaskKey(2),
+				},
+				{
+					Range: Range{
+						InclusiveMinTaskKey: persistence.NewImmediateTaskKey(4),
+						ExclusiveMaxTaskKey: persistence.NewImmediateTaskKey(6),
+					},
+					NextTaskKey: persistence.NewImmediateTaskKey(4),
+				},
+			},
+		},
+		{
+			name: "Case 5: [a,x,b,c,y,z] - Left range contains right range",
+			left: &GetTaskProgress{
+				Range: Range{
+					InclusiveMinTaskKey: persistence.NewImmediateTaskKey(1),
+					ExclusiveMaxTaskKey: persistence.NewImmediateTaskKey(6),
+				},
+				NextTaskKey: persistence.NewImmediateTaskKey(3),
+			},
+			right: &GetTaskProgress{
+				Range: Range{
+					InclusiveMinTaskKey: persistence.NewImmediateTaskKey(2),
+					ExclusiveMaxTaskKey: persistence.NewImmediateTaskKey(4),
+				},
```

Comment:
Is this a real possible case? We only merge if predicates match and there shouldn't be two separate slices with overlapping progress, right?